File: Sources/EMKECoordinator/PCMVoiceActivityDetector.swift
import Foundation

public enum PCMVoiceActivityEvent: Equatable, Sendable {
    case none
    case speechStarted
    case speechEnded
}

public enum PCMVoiceActivityDetectorError: Error, Equatable, Sendable {
    case invalidPCM16ByteCount
}

public struct PCMVoiceActivityDetector: Sendable {
    public let speechThreshold: Double
    public let silenceFrameLimit: Int

    public private(set) var isSpeaking = false
    private var consecutiveSilentFrames = 0

    public init(
        speechThreshold: Double = 0.015,
        silenceFrameLimit: Int = 30
    ) {
        precondition(speechThreshold >= 0)
        precondition(silenceFrameLimit > 0)
        self.speechThreshold = speechThreshold
        self.silenceFrameLimit = silenceFrameLimit
    }

    public mutating func observe(
        _ pcm16: Data
    ) throws -> PCMVoiceActivityEvent {
        guard pcm16.count.isMultiple(of: 2) else {
            throw PCMVoiceActivityDetectorError.invalidPCM16ByteCount
        }
        guard !pcm16.isEmpty else { return .none }

        let rms = pcm16.withUnsafeBytes { rawBuffer -> Double in
            let bytes = rawBuffer.bindMemory(to: UInt8.self)
            var squareSum = 0.0
            var index = 0
            while index < bytes.count {
                let bits = UInt16(bytes[index])
                    | UInt16(bytes[index + 1]) << 8
                let normalized = Double(Int16(bitPattern: bits)) / 32_768
                squareSum += normalized * normalized
                index += 2
            }
            return sqrt(squareSum / Double(bytes.count / 2))
        }

        if rms >= speechThreshold {
            consecutiveSilentFrames = 0
            if !isSpeaking {
                isSpeaking = true
                return .speechStarted
            }
            return .none
        }

        guard isSpeaking else { return .none }
        consecutiveSilentFrames += 1
        guard consecutiveSilentFrames >= silenceFrameLimit else {
            return .none
        }

        isSpeaking = false
        consecutiveSilentFrames = 0
        return .speechEnded
    }

    public mutating func reset() {
        isSpeaking = false
        consecutiveSilentFrames = 0
    }
}

File: Windows/src/EMKE.Routing/PcmVoiceActivityDetector.cs
namespace EMKE.Routing;

public enum PcmVoiceActivityEvent
{
    None,
    SpeechStarted,
    SpeechEnded,
}

public sealed class PcmVoiceActivityDetector
{
    private int _consecutiveSilentFrames;

    public PcmVoiceActivityDetector(
        double speechThreshold = 0.015,
        int silenceFrameLimit = 30)
    {
        if (!double.IsFinite(speechThreshold) || speechThreshold < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(speechThreshold));
        }

        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(silenceFrameLimit);

        SpeechThreshold = speechThreshold;
        SilenceFrameLimit = silenceFrameLimit;
    }

    public double SpeechThreshold { get; }

    public int SilenceFrameLimit { get; }

    public bool IsSpeaking { get; private set; }

    public PcmVoiceActivityEvent Observe(ReadOnlySpan<byte> pcm16)
    {
        if ((pcm16.Length & 1) != 0)
        {
            throw new ArgumentException(
                "PCM16 input must contain an even number of bytes.",
                nameof(pcm16));
        }

        if (pcm16.IsEmpty)
        {
            return PcmVoiceActivityEvent.None;
        }

        double squareSum = 0;
        for (int index = 0; index < pcm16.Length; index += 2)
        {
            ushort bits = (ushort)(pcm16[index] | (pcm16[index + 1] << 8));
            short sample = unchecked((short)bits);
            double normalized = sample / 32_768d;
            squareSum += normalized * normalized;
        }

        double rms = Math.Sqrt(squareSum / (pcm16.Length / 2d));
        if (rms >= SpeechThreshold)
        {
            _consecutiveSilentFrames = 0;
            if (!IsSpeaking)
            {
                IsSpeaking = true;
                return PcmVoiceActivityEvent.SpeechStarted;
            }

            return PcmVoiceActivityEvent.None;
        }

        if (!IsSpeaking)
        {
            return PcmVoiceActivityEvent.None;
        }

        _consecutiveSilentFrames++;
        if (_consecutiveSilentFrames < SilenceFrameLimit)
        {
            return PcmVoiceActivityEvent.None;
        }

        IsSpeaking = false;
        _consecutiveSilentFrames = 0;
        return PcmVoiceActivityEvent.SpeechEnded;
    }

    public void Reset()
    {
        IsSpeaking = false;
        _consecutiveSilentFrames = 0;
    }
}
